### PR TITLE
Enable react-hooks/exhaustive-deps eslint rules

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

As suggested in the [Rules of hooks](https://reactjs.org/docs/hooks-rules.html) guide in the official doc, it's highly recommended to enable the `react-hooks/exhaustive-deps` eslint rules to prevent us from writing fragile code.

There are currently a lot of violations to the rules, this PR doesn't intend to fix them now, but only emit warnings to prevent breaking backward compatibility.

I have to also upgrade `eslint` and `eslint-plugin-react-hooks` to support optional chaining.

## How has this been tested?

Run the `npm run lint-js` command.

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by noisysocks** [comment]
It looks like `react-hooks/exhaustive-deps` was added in https://github.com/WordPress/gutenberg/pull/14995 but then removed in https://github.com/WordPress/gutenberg/pull/14985. There was some discussion about it in https://github.com/WordPress/gutenberg/pull/14995#discussion_r275943117 and https://github.com/WordPress/gutenberg/pull/14985/files#r355242262.

Personally I'm a 👍 as I've found myself wanting this rule on several occasions.

@youknowriad: Do you remember what exceptions to this rule we were running into?

**Comment by noisysocks** [comment]
@kevin940726: Could you please add a CHANGELOG entry to `packages/eslint-plugin`? This change means that we need to increment the minor version field.

**Comment by kevin940726** [comment]
@noisysocks Ahh didn't know that! Good to know ❤️ 

It seems like when we're removing it we didn't really understand the usage of it since there isn't much documentation about it either. I think that's not applicable now, it does more good than harm, and enabling them only gives us warnings not errors.

Sure, I'll add an entry!

**Comment by youknowriad** [comment]
I think there are a lot of places where we explicitly rely on the fact that we don't want to run the "effect" when some variable changes. I don't know which ones though. So I'm not certain it's a good approach to just go ahead and add a warning for this. 

I'd rather we audit where the rule fails and see what impact applying the rule has. 

I know this just adds "warnings", but a warning on the output of eslint doesn't tell the whole story that potentially, this is not an issue.

**Comment by kevin940726** [comment]
@youknowriad 

Agreed with the "just add warnings" part, it definitely adds some noises, so I'd expect us to fix them all in the future.

> I think there are a lot of places where we explicitly rely on the fact that we don't want to run the "effect" when some variable changes.

Have to disagree with this though. As pointed by Dan in https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528, disabling it often leads to bugs and is usually a mistake. Furthermore, we have a big open source community with many contributors with different levels of skill, enabling the "official recommended" eslint rules not only helps us have consensus on coding style, but also helps contributors write resilient code. It also helps us reviewing the code too (https://github.com/WordPress/gutenberg/pull/23952#discussion_r478787502)!

I'd suggest to enable the rules, and if we did find a proper use case where eslint rules shouldn't be enabled, we can always disable it inline and add comments above it. If we believe it's a bug, we can even submit it to https://github.com/facebook/react/issues/14920 :)

**Comment by youknowriad** [comment]
Happy to be proved wrong here and I know Dan Abramov's opinion on this but IMO it's not black and white here. I'm certain that this rule is helpful in 80% of our current warnings but the audit will help make sure whether we can trust it 100%

**Comment by ZebulanStanphill** [comment]
I think if we want to explicitly leave a variable out of a dependency array, we need to make it clear that we're doing so intentionally via a lint-rule disable for that line and an associated comment explaining why it's okay to not include every dependency in that particular use-case.

It's also worth noting that for cases where we want to run something only once, we could simply introduce an explicitly titled `useEffectOnce` (a pattern I've seen in a couple React libraries already).

For those reasons, I am completely in favor of introducing this ESLint rule.

File: packages/compose/src/hooks/use-shallow-compare-effect/index.js
Comment (ellatrix) [inline packages/compose/src/hooks/use-shallow-compare-effect/index.js:23]: How does this work? When setting `ref.current`, there's no guarantee that the component will re-render, and therefore no guarantee that `useEffect` and the callback will be called.

**Comment by ellatrix** [comment]
Not sure if I agree with adding refs to the dependencies. https://github.com/WordPress/gutenberg/pull/27080#discussion_r530229783

> A ref by definition never changes, so it is pointless to add to the dependencies. It gives the false sense that it will be reevaluated when the `ref.current` changes, which is not the case.

Also adding `ref.current` to the dependency array is not right, because setting `.current` does not cause a component to re-render.

I don't have a strong opinion on this though.

File: packages/compose/src/hooks/use-shallow-compare-effect/index.js
Comment (kevin940726) [reply packages/compose/src/hooks/use-shallow-compare-effect/index.js:23]: It's an escape hatch, so not intuitive at all 😞.

How it works:
1. When we receive a new `deps` in the render phase, we compare it with the last value we've seen (`ref.current`).
2. If they are shallowly equal, we don't change `ref.current`, so that the `useEffect` below won't run.
3. If they are not shallow equal, we set `ref.current` to the new `deps`. Here we assume that if two `deps` are not shallowly equal, then they are also not referential equal.
4. If `ref.current` changes, the effect will run.

So basically, it's not that `ref.current` triggers the re-render, but the re-render triggers the setting of the new `ref.current`, and therefore triggers the effect to run.

It's essentially how [`use-deep-compare-effect`](https://github.com/kentcdodds/use-deep-compare-effect/blob/master/src/index.ts) is implemented, but with shallow equality rather than deep equality.

FWIW though, it seems like setting `ref.current` in the render phase is now considered a anti-pattern. I don't think we have to change it now though since this hook is mostly for backward-compatibility anyway 🤔 .

**Comment by ciampo** [comment]
Hey all, just checking the status of this PR — are there any plans to resume work on it?

I personally believe that the `exhaustive-deps` is one of the most important `eslint` rules, and that most of the times specifying all deps is the correct behaviour (and not specifying all deps can lead to hard-to-debug bugs).

For the few times where we need to specify fewer deps, adding a comment to disable the rule and an explanation of the whys should be enough.

**Comment by kevin940726** [comment]
@ciampo Hi 👋 ! I'm glad you asked! I 100% agree with you here.

The initial strategy was to fix everything to follow the new rule, but after some iterations, I think we'll never be able to catch up on the new code added to this repo.

I'd suggest enabling this rule under the "warning" severity level now so that people can be aware of this new best practice and act accordingly. We can disable outputting warnings in CLI with the `--quite` flag.

WDYT? Feel free to take this over though, since I won't have time to work on this for the next couple of weeks.

**Comment by ciampo** [comment]
> The initial strategy was to fix everything to follow the new rule, but after some iterations, I think we'll never be able to catch up on the new code added to this repo.

Have we assessed how many errors/warning we'd be introducing by enabling this rule? It could help us to make a better informed decision on what do to next.

> I'd suggest enabling this rule under the "warning" severity level now so that people can be aware of this new best practice and act accordingly. We can disable outputting warnings in CLI with the `--quiet` flag.
> WDYT?

What would that cause? The warning would be only displayed in a IDE, but not in the console ?

I think the ideal solution for me would be to somehow prevent a new PR from introducing more instances of this warning. And then we can slowly work our way down to 0.

Although I'm probably not the better suited here to make such a decision. I'm also going to ping "refactor extraordinaire" @tyxla in case he has an advice or previous knowledge around this topic.

> Feel free to take this over though, since I won't have time to work on this for the next couple of weeks.

I can't guarantee anything at the moment, but let's see first what's the outcome of this conversation

**Comment by tyxla** [comment]
Like for any other big project, I think changes like this should be done gradually and incrementally. 

As a first step, I'd see the `react-hooks/exhaustive-deps` rule being enabled as a warning (`1`). Yes, it would introduce a bunch of new warnings, but it would also help against introducing new ones. It wouldn't force against introducing new ones, but it would still be an improvement IMHO. It would also inform the rest of the folks working on the same code and perhaps in some instances they might drive-by fix those instances. 

Simultaneously, we could work on reducing the warnings one by one. Once we're close (or done), we could enable the rule as an error. 

I'd just suggest keeping an eye on the goal to reduce them to zero - I've seen some instances where we intentionally want to exclude items that look like dependencies but shouldn't be. In those cases, we might want to intentionally disable the rule inline.

**Comment by ciampo** [comment]
@tyxla 's advice seems very sensible to me — should we go ahead and:

1. Add the rule as a warning
2. Have a tracking issue where we list all the warnings. Progressively, we aim at removing all warnings:
   - by explicitly disabling the rule and adding a comment explaining the exception; or,
   - by adding all deps (and thus fixing the warning)
3. Once all warnings have gone, switch the linter rule from warning to error

**Comment by kevin940726** [comment]
> I've seen some instances where we intentionally want to exclude items that look like dependencies but shouldn't be. In those cases, we might want to intentionally disable the rule inline.

Worth mentioning that it should be really rare to have to disable the lint rule explicitly. There are often better alternatives we can leverage in those cases. See this awesome [summary](https://github.com/facebook/react/issues/14920#issuecomment-471070149) from React core team for more info.

Another thing we should take into account is that we're also using deps arrays in some custom hooks (`useSelect`, `useRefEffect`, etc). We should also enable them to be checked by eslint via the [advanced configuration](https://www.npmjs.com/package/eslint-plugin-react-hooks#advanced-configuration). (Note that adding deps to custom hooks is discouraged, whether or not the existing APIs should be changed is debatable, but that's not the point of this issue.)

> Have a tracking issue where we list all the warnings.

This is great. We can also provide some common ways to fix it in the issue so that contributors can help fix them as well (might be a good chance to label it `good-first-issue`).

**Comment by ciampo** [comment]
I agree with what @kevin940726 mentioned re. trying better alternatives to rules exceptions and custom hooks. I guess those cases will need to be assessed individually as we'll slowly make our way through the codebase removing the warnings.

> We can also provide some common ways to fix it in the issue so that contributors can help fix them as well (might be a good chance to label it good-first-issue).

This is also a good point!

In terms of capacity, would anyone be able to start working on introducing the rule and starting a tracking issue?

**Comment by kevin940726** [comment]
Just want to reference the new(beta) React [doc](https://beta.reactjs.org/learn/removing-effect-dependencies#to-change-the-dependencies-change-the-code) here to better explain why it's generally a bad idea to disable this lint rule.

**Comment by ciampo** [comment]
That's very useful, thank you!

We've been doing some work in the components package (https://github.com/WordPress/gutenberg/pull/41166 cc @chad1008 ), and we've been introducing `// disable` comments only:
 - when adding back all the dependencies would cause regressions in the component's behaviour, and therefore a deeper rewrite was necessary
 - in `.native` components, since we'd rather let the native team handle the migration

We should not be too far from being able to reintroduce this rule, at least in the components package 🤞

**Review by scruffian — APPROVED** [review]
This rule would have helped us to catch this bug: https://github.com/WordPress/gutenberg/issues/47881, which almost made it in to WordPress 6.2.

**Comment by getdave** [comment]
👋 Hi folks. I was pleased to find this PR. Just to flag a real world use case as evidence for how much we need this rule to be on.

There was a bug in the Nav block which would have been a regression for WP 6.2. Luckily I caught it by chance, but we couldn't work out why the bug was happening. 

Turns out [the `useEffect` deps had got out of sync ](https://github.com/WordPress/gutenberg/pull/47912)and this caused the problem.

If we had this warning on we could have caught the bug earlier with no need for automated testing at all.

I've taken the liberty of rebasing this PR (please check I didn't break it) in the hope that we can look to merge it.

I would vote for [enabling it as a warning](https://github.com/WordPress/gutenberg/pull/24914#issuecomment-1032374149) as suggested by @ciampo above. We can then work towards zero.

**Comment by kevin940726** [comment]
Just FYI that the rule allows custom hooks to be checked as well. Some internal hooks we use like `useSelect` or `useRefEffect` can also benefit from it. This can be enabled in a different PR though.

**Comment by tyxla** [comment]
Noting that it is currently enabled as `error`, but only for the components package:

https://github.com/WordPress/gutenberg/blob/26d2f37bfd019772f34cd12efc5582e3633dc080/.eslintrc.js#L355

I'm happy to introduce this as a warning in the eslint plugin since we have great proof that it helps 👍

File: packages/eslint-plugin/configs/react.js
Comment (tyxla) [inline packages/eslint-plugin/configs/react.js:37]: Nit: we could keep the list alphabetized. Sometimes that helps when reading the code.

**Review by tyxla — APPROVED** [review]

**Comment by getdave** [comment]
Shall we just merge this now then? Once done we can drop a note in Core Slack to advise everyone not to panic and how they can go about resolving errors.

File: packages/eslint-plugin/configs/react.js
Comment (getdave) [reply packages/eslint-plugin/configs/react.js:37]: I'll fix this now.

Comment (getdave) [reply packages/eslint-plugin/configs/react.js:37]: ✅

**Comment by tyxla** [comment]
> Shall we just merge this now then? Once done we can drop a note in Core Slack to advise everyone not to panic and how they can go about resolving errors.

I don't see why not. This will only land in the next release of the eslint-plugin package and it does have a CHANGELOG entry, plus it's only a warning, so it will not be blocking any work. Of course, a note in Slack or a dev note would be helpful to inform everyone interested.

**Comment by tyxla** [comment]
Just another note - we should aim to fix any new warnings of that type in the Gutenberg codebase. The idea is not to clutter it with warnings 😉

**Comment by ciampo** [comment]
With the changes from this PR, would the rule still error in the `@wordpress/components` package?

---

For the strategy towards "zero warnings", I suggest the approach that we took for the components package:
- focus on one package at a time
- remove all warnings, either by satisfying eslint demants, or by ignoring them when a rewrite would be complicated (while also adding a code comment explaining the reason for the disable)
- once all warnings are gone, change the rule for that specific package from `warn` to `error` to avoid re-introducing more errors
- pick a new package, repeat

**Comment by tyxla** [comment]
> With the changes from this PR, would the rule still error in the @wordpress/components package?

Yes. This is changing the default rules, but those are overridden by the project ESLint settings, where the rule is set to error for the components package.

**Comment by getdave** [comment]
What's going on with the failure on Static Analysis here?

**Comment by tyxla** [comment]
Could it be because of the prettier dependency change? Is that even intentional or it got here during a bad rebase?

**Comment by Mamaduka** [comment]
I disabled 'auto-merge' since lint checks are failing, and there's still an ongoing discussion.

@kevin940726, I remember you explored the idea of ignoring methods returned by `useDispatch` from the rules. Is it still a lot of work to customize this behavior?

**Comment by kevin940726** [comment]
> I remember you explored the idea of ignoring methods returned by `useDispatch` from the rules. Is it still a lot of work to customize this behavior?

I don't remember that 😅 . I think we can just add them to the dependencies array, it might be the simplest way. This is how `react-redux`'s `useDispatch` does it too.

File: packages/eslint-plugin/package.json
Comment (getdave) [inline packages/eslint-plugin/package.json]:
```suggestion
```

@tyxla Let's remove this

**Comment by getdave** [comment]
If the lint still fails then it might be we need to rebuild `package-lock.json` and commit that?

**Comment by tyxla** [comment]
> Could it be because of the prettier dependency change? Is that even intentional or it got here during a bad rebase?

Seems like that was the cause.

File: packages/eslint-plugin/CHANGELOG.md
Comment (gziolo) [inline packages/eslint-plugin/CHANGELOG.md:174]: This changelog entry is misplaced. It should be moved to `## Unreleased` section.

We will have to rethink the way changelogs are maintained, as it's too easy to run accidentaly into this type of situation when merging the latest changes from `trunk`.

Comment (ciampo) [reply packages/eslint-plugin/CHANGELOG.md:174]: > We will have to rethink the way changelogs are maintained,

100%. I was thinking if we could semi-automate the changelog by asking the PR author to simply state:
- which category this falls under (bug fix, enhancements, breaking change...)
- message

The rest can be performed by the automated action at merge time, including picking the right CHANGELOG file and adding the message under the right version (or unreleased) ?

Comment (gziolo) [reply packages/eslint-plugin/CHANGELOG.md:174]: I was looking at [Jetpack monorepo](https://github.com/Automattic/jetpack) the other day, and they use an interesting system.

They keep a `changelog` folder for every package:

Screenshot 2023-02-13 at 09 33 51

Every time a changelog entry is warranted (I don't know the rules), they create a new file:

Screenshot 2023-02-13 at 09 33 40

Those files are read and deleted at publishing time, which completely removes the risk of merge conflicts of any type.

Example PR from Jetpack with changelog entries included: https://github.com/Automattic/jetpack/pull/28710.

Comment (ciampo) [reply packages/eslint-plugin/CHANGELOG.md:174]: That's interesting!

It could be worth chatting to whoever implemented / maintains that system to understand pros and cons, and if it could work well for our use case. For example, I see that they have a CI action that replies to the PR with a checklist, suggesting that a CHANGELOG entry is a mandatory requirement for all PRs?

**Pull request merged**
